presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PreferredProperties.java

```diff
@@ -96,6 +100,70 @@ public static PreferredProperties local(List<? extends LocalProperty<Symbol>> lo
         return localProperties;
     }
 
+    public static PreferredProperties derivePreferences(PreferredProperties parentProperties,
+            Set<Symbol> partitioningColumns,
+            List<LocalProperty<Symbol>> localProperties)
+    {
+        // Build identity translations
+        ImmutableMap.Builder<Symbol, Symbol> translations = ImmutableMap.builder();
+        partitioningColumns.forEach(sym -> translations.put(sym, sym));
+        return derivePreferences(parentProperties, translations.build(), Optional.of(partitioningColumns), Optional.empty(), localProperties);
+    }
+
+    public static PreferredProperties derivePreferences(PreferredProperties parentProperties,
+            Map<Symbol, Symbol> translations,
+            Optional<Set<Symbol>> partitioningColumns,
+            Optional<List<Symbol>> hashingColumns,
+            List<LocalProperty<Symbol>> localProperties)
+    {
+        List<LocalProperty<Symbol>> parentLocalProperties = LocalProperties.translate(parentProperties.getLocalProperties(), column -> Optional.ofNullable(translations.get(column)));
+
+        // if the child plan can first meet our requirements and then by our parent's,
+        // it can satisfy both in one shot
+        List<LocalProperty<Symbol>> local = ImmutableList.<LocalProperty<Symbol>>builder()
+                .addAll(localProperties)
+                .addAll(parentLocalProperties)
+                .build();
+
+        // Check we need to be hash partitioned
+        if (hashingColumns.isPresent() && partitioningColumns.isPresent() && partitioningColumns.get().equals(ImmutableSet.copyOf(hashingColumns.get()))) {
```

Comment:
This doesn't look right to me. If hashing columns is present, it must always be equal to partitioning columns. We might want to add a precondition for that and change this check to:

```
if (hashingColumns.isPresent()) { 
    return ...
}
```